src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,205 @@
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (sender tx-sender)
+    )
+
+    (asserts! (> amount u0) (err ERR_AMOUNT_NOT_POSITIVE))
+    (asserts! (>= cur-principal-votes amount) (err ERR_INSUFFICIENT_FUNDS))
+
+    (unwrap! (as-contract (stx-transfer? amount tx-sender sender)) (err ERR_STX_TRANSFER))
+    (unwrap! (as-contract (ft-transfer? cost-vote-token amount sender tx-sender))
+        (err ERR_FT_TRANSFER))
+
+    (map-set proposal-votes { proposal-id: proposal-id } { votes: (- cur-votes amount) })
+    (map-set principal-proposal-votes { address: tx-sender, proposal-id: proposal-id }
+                                        { votes: (- cur-principal-votes amount) })
+    (ok true))
+)
+
+;; Miner veto
+(define-public (veto (proposal-id uint))
+    (let (
+        (cur-vetos (default-to u0 (get vetos (map-get? proposal-vetos { proposal-id: proposal-id }))))
+        (expiration-block-height (get expiration-block-height (unwrap! (map-get? proposals {
+            proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL))))
+        (vetoed (default-to false (get vetoed (map-get? exercised-veto { proposal-id: proposal-id,
+                                                                         veto-height: block-height }))))
+        (last-miner (unwrap! (get-block-info? miner-address (- block-height u1))
+            (err ERR_FETCHING_BLOCK_INFO)))
+    )
+    
+    (asserts! (not vetoed) (err ERR_ALREADY_VETOED))
+    (asserts! (is-eq contract-caller last-miner) (err ERR_NOT_LAST_MINER))
+    (asserts! (> burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+    (asserts! (is-none (map-get? proposal-confirmed-id { proposal-id: proposal-id }))
+        (err ERR_PROPOSAL_CONFIRMED))
+
+    (map-set proposal-vetos { proposal-id: proposal-id } { vetos: (+ u1 cur-vetos) })
+    (map-set exercised-veto { proposal-id: proposal-id, veto-height: block-height }
+                            { vetoed: true })
+    (ok true))
+)
+
+;; Confirm proposal has reached required vote count
+(define-public (confirm (proposal-id uint))
+    (let (
+        (votes (default-to u0 (get votes (map-get? proposal-votes { proposal-id: proposal-id }))))
+        (vetos (default-to u0 (get vetos (map-get? proposal-vetos { proposal-id: proposal-id }))))
+        (proposal (unwrap! (map-get? proposals { proposal-id: proposal-id }) (err ERR_NO_SUCH_PROPOSAL)))
+        (confirmed-count (var-get confirmed-proposal-count))
+    )
+    (let (
+        (expiration-block-height (get expiration-block-height proposal))
+    )
+    (let (
+        (blocks-since-proposal (- block-height (- expiration-block-height u2016)))
+    )
+
+    (asserts! (< burn-block-height expiration-block-height) (err ERR_PROPOSAL_EXPIRED))
+    (asserts! (>= (/ (* votes u100) stx-liquid-supply) u20) (err ERR_INSUFFICIENT_VOTES))
```

Comment:
You might want to make these constants, and document them.  Is this documentation correct?

* "A proposal must voted for by at least 20% of the total liquid STX"
* "A proposal must be vetoed by no more than 20% of the mined blocks since the proposal was created"